core/api/usermanager/v1/usermanager_proto/usermanager.proto

```diff
@@ -0,0 +1,156 @@
+
+package keytransparency.usermanager.v1;
+
+import "google/api/annotations.proto";
+import "google/protobuf/field_mask.proto";
+import "proto/keytransparency_v1_proto/keytransparency_v1_proto.proto";
+import "proto/keymaster_proto/keymaster_proto.proto";
+
+// GetKeySetsRequest requests the keyset of a domain_id/app_id
+message GetKeySetRequest {
+   // domain_id identifies the domain.
+   string domain_id = 1;
+   // app_id identifies the application.
+   string app_id = 2;
+}
+
+// CreateUserRequest specifies the information with which a new user should be initialized.
+// New users will be signed with the current active key.
+// It is the responsibility of authorized callers to verify that domain_id/app_id/user_id is correct.
+message CreateUserRequest {
+  // user is the user to create.
+  keytransparency.v1.proto.User user = 2;
+  // add_signing_keys specifies whether to add this service's signing keys to the set of authorized_keys.
+  // This must be set to true if any further operations from this API are meant to succeed.
+  // If set to false, there must be at least one key in authorized_keys.
+  bool add_signing_keys = 7;
+}
+
+// UpdateUserRequest sets the data field for the user.
+// The user must have the service's current signing key in its list of
+// authorized_keys in order to succeed.
+message UpdateUserRequest {
+  // user contains data which will be applied to the user.
+  keytransparency.v1.proto.User user = 2;
+  // update_mask specifies which fields of user to update.
+  // For example: "data" or "authorized_keys"
+  google.protobuf.FieldMask update_mask = 3;
+}
+
+// BatchCreateUserRequest creates multiple users all at once.
+message BatchCreateUserRequest {
+   // domain_id identifies the domain.
+   string domain_id = 1;
+   // app_id identifies the application.
+   string app_id = 2;
+  // users is the set of users to create.
+  repeated keytransparency.v1.proto.User users = 3;
+  // add_signing_keys specifies whether to add this service's signing_keys to the set of authorized_keys.
+  // This must be set to true if any further operations from this API are meant to succeed.
+  // If set to false, there must be at least one key in authorized_keys.
+  bool add_signing_keys = 4;
+}
+
+// BatchCreateUserRequest creates multiple users at once.
+message BatchCreateUserResponse {
+  // users returns the list of created users.
+  repeated keytransparency.v1.proto.User users = 2;
+}
+
+// BatchUpdateUserRequest creates multiple users all at once.
+message BatchUpdateUserRequest {
+   // domain_id identifies the domain.
+   string domain_id = 1;
+   // app_id identifies the application.
+   string app_id = 2;
+  // users is the set of users to create.
+  repeated keytransparency.v1.proto.User users = 3;
+  // update_mask specifies which fields of user to update.
+  // For example: "data" or "authorized_keys"
+  google.protobuf.FieldMask update_mask = 4;
```

Comment:
I don't recall seeing a BatchUpdate before, so there isn't much I can draw from, but since all updates have to use the same mask, do you think this method will end up being useful enough to be here? Are you foreseeing some sort of "batch backfill" that touches the same fields throughout large numbers of users?

Another question would be: does the BatchUpdate require a List method in order to be useful, or would the client be aware of the user space without it?

Reply:
I foresee wanting to update all the authorized_keys in a consistent manner such as
- Remove Key X
- Add Key Y

But that doesn't work very well with the update_mask on a list does it?

Your list question is very astute.  The Key Transparency API does not have a list option on purpose in order to not help spammers.  This API, however, could keep track of the IDs that it created on it's own.